File: lib/engine/funding.ts
// ============================================================================
// Aegis Vault — Funding Rate Calculation & Application
// Implements hourly funding rate with ±75bps cap.
// ============================================================================

import {
  type Position,
  Side,
  MAX_FUNDING_RATE_BPS,
  FUNDING_FACTOR,
} from './types';

/**
 * Compute the funding rate given mark and index prices.
 * Formula: (mark - index) * FUNDING_FACTOR / index
 * Result is in basis points, capped at ±75 bps per hour.
 *
 * Positive rate: longs pay shorts.
 * Negative rate: shorts pay longs.
 */
export function computeFundingRate(
  markPrice: number,
  indexPrice: number
): number {
  if (indexPrice <= 0) throw new Error('Index price must be positive');

  const rawRate =
    ((markPrice - indexPrice) * FUNDING_FACTOR) / indexPrice;

  // Cap at ±75 bps
  return Math.max(
    -MAX_FUNDING_RATE_BPS,
    Math.min(MAX_FUNDING_RATE_BPS, rawRate)
  );
}

/**
 * Apply a funding rate to a position.
 * Longs pay when rate > 0, shorts pay when rate < 0.
 * Payment = position.size * position.entryPrice * fundingRate / FUNDING_FACTOR
 *
 * Mutates position.unrealizedPnl.
 * Returns the funding payment amount (positive = paid out, negative = received).
 */
export function applyFunding(
  position: Position,
  fundingRate: number
): number {
  const notional = position.size * position.entryPrice;
  const payment = (notional * fundingRate) / FUNDING_FACTOR;

  if (position.side === Side.Long) {
    // Longs pay when rate is positive (mark > index)
    position.unrealizedPnl -= payment;
    return payment;
  } else {
    // Shorts receive when rate is positive (mark > index)
    position.unrealizedPnl += payment;
    return -payment;
  }
}

File: engine/src/funding.rs
use crate::types::{Position, Side, FUNDING_FACTOR, MAX_FUNDING_RATE_BPS};

pub fn compute_funding_rate(mark_price: f64, index_price: f64) -> Result<f64, &'static str> {
    if index_price <= 0.0 {
        return Err("Index price must be positive");
    }

    let raw_rate = ((mark_price - index_price) * FUNDING_FACTOR) / index_price;

    Ok(raw_rate.clamp(-MAX_FUNDING_RATE_BPS, MAX_FUNDING_RATE_BPS))
}

pub fn apply_funding(position: &mut Position, funding_rate: f64) -> f64 {
    let notional = position.size * position.entry_price;
    let payment = (notional * funding_rate) / FUNDING_FACTOR;

    if position.side == Side::Long {
        position.unrealized_pnl -= payment;
        payment
    } else {
        position.unrealized_pnl += payment;
        -payment
    }
}
